### DISCUSSION TITLE
How to resolve unknown import symbol for flask_pydantic.validate

### DISCUSSION THREAD

**Discussion by acartine** [microsoft/pyright#4733] [Q&A] [answered]
I've been following this guide
https://github.com/microsoft/pyright/blob/main/docs/getting-started.md

I got to step 5 and (thankfully) the only error I got was:

```
/home/acartine/code/nifty/src/app.py
  /home/acartine/code/nifty/src/app.py:5:6 - error: Stub file not found for "flask_pydantic" (reportMissingTypeStubs)
```
  
 so i created the stub
 
```
 (nifty) ➜  nifty git:(hotlink-previews) ✗ pyright --createstub flask_pydantic 
Loading configuration file at /home/acartine/code/nifty/pyrightconfig.json
stubPath /home/acartine/code/nifty/typings is not a valid directory.
Type stub was created for 'flask_pydantic'
```

then i get

```
(nifty) ➜  nifty git:(hotlink-previews) ✗ pyright                             
Loading configuration file at /home/acartine/code/nifty/pyrightconfig.json
Searching for source files
Found 31 source files
pyright 1.1.296
/home/acartine/code/nifty/src/app.py
  /home/acartine/code/nifty/src/app.py:5:28 - error: "validate" is unknown import symbol (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 3.03sec
```

I am using virtual environments, and i noticed that all the things I tried to specify the venv didn't seem to have an effect.  Could that be related?  For example, my virtual environment is located at

`/home/acartine/.local/share/virtualenvs/nifty-D6-AHl3j`

when i specify -`-venv-path /home/acartine/.local/share/virtualenvs` or `/home/acartine/.local/share/virtualenvs/nifty-D6-AHl3j` the debug logs show that pyright is still searching system paths:

```
Finding python search paths
  Executing interpreter: 'python3'
  Skipping '/usr/lib/python310.zip' because it is not a valid directory
  Received 3 paths from interpreter
    /usr/lib/python3.10
    /usr/lib/python3.10/lib-dynload
    /home/acartine/.local/share/virtualenvs/nifty-D6-AHl3j/lib/python3.10/site-packages
  Looking in python search path '/usr/lib/python3.10'
  Attempting to resolve stub package using root path '/usr/lib/python3.10'
  Attempting to resolve using root path '/usr/lib/python3.10'
  Looking in python search path '/usr/lib/python3.10/lib-dynload'
  Attempting to resolve stub package using root path '/usr/lib/python3.10/lib-dynload'
  Attempting to resolve using root path '/usr/lib/python3.10/lib-dynload'
  Looking in python search path '/home/acartine/.local/share/virtualenvs/nifty-D6-AHl3j/lib/python3.10/site-packages'
...
```

would really like to understand and resolve issues like this as pyright has already helped me find numerous bugs!

**Comment by erictraut** [comment, marked as answer]
The `flask_pydantic` package appears to have inline type annotations within its own code. However, it is not marked as "py.typed" (i.e. it doesn't have a "py.typed" marker file indicating that it has inlined types). You may want to reach out to the maintainers of the library to encourage them to add this marker file. For more details, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md).

Make sure you're using the latest version of flask_pydantic. It should be "0.11.0".

In the meantime, you have two options:
1. Enable the `useLibraryCodeForTypes` option in pyright. This tells it that it should infer type information from the library code even if the library is not marked "py.typed".
2. Generate your own local stubs, as you appear to have done. This will create a "first draft" of the stubs. You'll likely need to massage them somewhat to fix problems that the stub generator was unable to resolve. The `"validate" is unknown` error is a symptom of this. You can fix this by opening the generated stub file `flask_pydantic/__init__.pyi` and changing the line `from .core import validate` to `from .core import validate as validate`. This tells pyright (and other type checkers) that the top-level module is intended to import and re-export the `validate` symbol (which is defined in the `core` submodule) from the top-level namespace.

**Reply by acartine** [reply]
https://github.com/bauerji/flask-pydantic/issues/65

in the meantime have modified the generated stub file as suggested.  Thank you!

**Reply by acartine** [reply]
i can create a different discussion for this, but it's along the same vein.  when i add `  "reportUnknownParameterType": true,` i get a a bunch of errors from `redis-py` module.  Specifically, that `Redis[Unknown]` requires a type, either `str` or `bytes`.  And sure enough when i look at the stub in [`redis-stubs` ](https://github.com/python/typeshed/blob/main/stubs/redis/redis/client.pyi#L79) it shows the parameter type (corresponding to the `decode-responses` argument I presume).

so i went and typed all my code accordingly, and when I ran the code I got this error:

```
File "/home/acartine/code/nifty/src/nifty_common/redis_helpers.py", line 14, in <module>
    def get_redis(redis_type: RedisType) -> Redis[str]:
  File "/usr/lib/python3.10/typing.py", line 312, in inner
    return func(*args, **kwds)
  File "/usr/lib/python3.10/typing.py", line 1345, in __class_getitem__
    _check_generic(cls, params, len(cls.__parameters__))
  File "/home/acartine/.local/share/virtualenvs/nifty-D6-AHl3j/lib/python3.10/site-packages/typing_extensions.py", line 97, in _check_generic
    raise TypeError(f"{cls} is not a generic class")
TypeError: <class 'redis.client.Redis'> is not a generic class
```
when i look at the [source] (https://github.com/redis/redis-py/blob/master/redis/client.py#L845) there is no such parameter.  I've confirmed locally they my stub is at 4.5.1.4, the package at 4.5.1 (where i can also see that the executable source does not parameterize the Redis class).

So I'm sort of assuming that something is broken there - is it?  Or am I doing something else stupid?

If something is broken, how do i "override" the typings?  I generated my own stubs "temporarily" and it was quite a complicated mess trying to figure out how to clean it up.

**Reply by erictraut** [reply]
Apparently, the type stub deviates from the implementation. The stub indicates that the class is generic, but the implementation is not. I'd consider that a bug in the type stub. You may want to report it to the authors.

As a workaround, you can use `from __future__ import annotations` in your code or enclose the errant annotations in quotes (as in `"Redis[str]"`). Either of these options will cause the Python interpreter to not evaluate the annotation expression, so you can avoid the runtime exception.

**Reply by hmc-cs-mdrissi** [reply]
https://mypy.readthedocs.io/en/stable/runtime_troubles.html#using-classes-that-are-generic-in-stubs-but-not-at-runtime describes this problem. A number of stubs often lie about genericness as alternative of being non-generic would force a lot of Any's to be introduced when class implementation behaves generically, but doesn't actually use Generic[T] at runtime.

Typeshed commonly does this and while stub is lying I think being truthful to runtime would make a lot stubs much weaker and forced to fallback to use Any a lot. 

I think pandas stub also lie about this a good deal (pd.Dataframe is generic in stub but never at runtime) and even numpy depending on which python version you were on lied about this for very basic types (arrays). Numpy arrays are only generic at runtime >3.9, but stubs make it appear generic on all versions.

**Reply by acartine** [reply]
amazing, thank you both. super helpful.  sorry my google skills did not surface that mypy link.  :-(
